Update the Value section to reflect the code

polinavino opened the pull request:
Describe the Val class in the formal spec, and explain the mechanism by which the spec defines both the Mary and Allegra eras

newhoggy committed:
Run ormolu in Github Actions instead

Commit:
Initial work on Alonzo UTXOW property tests

lehins committed:
Continue work on generators for testing UTXOW:

    * Add Arbitrary instance for `KeyPair kd crypto`

    * Generators for `Timelocks`

    * Generation of plutus scripts, both valid and invalid

    * Utilize `RWST` for keeping track of generators state

    * Datums generation

    * Collateral generation

    * Generation of random ExUnits

polinavino committed:
value update to align with code

polinavino committed:
value update to align with code

WhatisRT commented on shelley-ma/formal-spec/introduction.tex:
```diff
@@ -1,12 +1,69 @@
+replaces the $\Coin$ type with a type the terms of which are bundles of arbitrary assets
+(user-defined ones, as well as lovelace), which we call $\Value$.
+
+In this specification, we represent all the places where the $\Coin$ type is used in Allegra,
```
Instead of calling it `ValClass`, we could also call it something like `LedgerMonoid` and define it the way these things are usually defined in mathematics. I.e. something along the lines of: "A `Ledger Monoid` is a monoid `V` together with functions ... such that ...`. We're generally trying to avoid programming jargon in the specs, so I think it would be nice to not use terms like `Class` or `instantiate`.

polinavino replied:
I will avoid using Class and Instantiate , but I think using the term LedgerMonoid is more confusing than saying something like "we define an algebraic structure Val, which is a monoid with additional functions". In that, I am pretty set on keeping the Val name as it is used in the code (perhaps typeset nicer than it currently is)

WhatisRT replied:
I don't particularly care about the name, but I think it's nice to have something descriptive. `Val` doesn't really tell you what operations it supports or what it's used for. I wanted to at least have `Monoid` in there so that the reader immediately knows that it's supposed to be a monoid with some extra stuff. Maybe `Val` is also not the best name for the class in the implementation?

WhatisRT replied:
Coming back to this, `Val` doesn't behave the same grammatically as other algebraic structures. You can say: "Let V be a vector space", but you can't say "Let V be a Val". I'd prefer it a lot if the name would fit in this sentence. Any of the following do: Ledger monoid, Value monoid, Accounting type, Coin-like type, token-bundle type, possibly something else. But having an algebraic structure that doesn't fit in the above sentence is awkward.

WhatisRT commented on shelley-ma/formal-spec/introduction.tex:
```diff
@@ -1,12 +1,69 @@
+abstraction representing any type that has certain functions defined on its terms,
+such as addition of two terms,
+a size function, etc. (see Section \ref{sec:value} for details). We say that we
+\emph{instantiate} $\ValClass$ with a specific type (eg. $\Coin$) when we interpretet all
```
TYPO: interpretet

WhatisRT commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,77 +1,225 @@
+  \emph{Helper function}
+  %
+  \begin{align*}
+    \mathsf{ValType} &\in \ValClass  \\
```
Why are there two definitions of `ValType`? Also, `ValType \in \ValClass` isn't really correct

WhatisRT replied:
This also has the wrong alignment in the PDF output

polinavino replied:
do you have any idea how to better specify the relation between `ValType` and `ValClass`?

WhatisRT replied:
See the comment above about the more traditional way of defining it. For a type class, it could either be a parameter of the class or a member of it. In the former case it's probably best to mention in in the prose and call it an `Abstract type`, and in the latter we could just say `ValType \in \Type` in this figure.

WhatisRT commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,77 +1,225 @@
+      & \text{Create a $\mathsf{ValType}$ term containing only this amount of Ada}
+      \nextdef
+      %
+      & \fun{adaOnly} ~\in~ \ValClass \to \Bool \\
```
The type should be ValType instead of ValClass here

WhatisRT commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,77 +1,225 @@
+      & \fun{adaOnly} ~\in~ \ValClass \to \Bool \\
+      & \fun{adaOnly}~v =
+        \begin{cases}
+          \True & \fun{inject} \circ \fun{coin} = \fun{id}_{\ValClass} \\
```
This condition does not depend on `v` and is either always or never true. Both sides need to be applied to `v`, i.e. `inject (coin v) = v`.

WhatisRT commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,77 +1,225 @@
+      \nextdef
+      %
+      & \fun{isZero} ~\in~ \mathsf{ValType} \to \Bool\\
+      & \fun{isZero}~v ~+~ v~=~\fun{zero} \\
```
This `+` should be an `=`

WhatisRT commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,77 +1,225 @@
-        c & \var{pid} = \mathsf{adaPolicy}~\text{and}~\var{aname} = \mathsf{adaName} \\
-        0 & \text{otherwise}
-      \end{cases}
+    & \fun{pids} ~\in~ \Value \to \powerset{\PolicyID} \\
```
This function should be part of the definition of ValClass

WhatisRT replied:
Also, because `Coin` is now separated out in `Value`, `supp` isn't well-defined anymore. I think for the spec we should keep the old definition of `Value`, and mention that we use something else in the implementation.

polinavino replied:
the reason `pids` (renamed to `policies` to match the code) and `anames` is not in `ValClass` is that it's not defined for Coin, nor does it need to be. These functions are not used anywhere on Coin.

WhatisRT replied:
Oh, I see, `minted` is the function we use.

WhatisRT commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,77 +1,225 @@
+    & \fun{pids}~v ~=~ \supp{~v}
+    \nextdef
+    %
+    & \fun{anames} ~\in~ \Value \to \powerset{\PolicyID} \\
```
Should also be part of ValClass

WhatisRT commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -80,46 +228,111 @@ \subsection*{Representing Multi-asset Types and Values}
-  \[\fun{inject} \circ \fun{coin}\]
+\begin{figure*}[t!]
+  \begin{align*}
+    & (c, v)~+~(d, w) =& (c+d, ~\{~ pid\mapsto_0 (aname\mapsto_0 (v~\var{pid}~\var{aname})~+~(w~\var{pid}~\var{aname}) ~\vert~ \\
```
This is really difficult to read (in the PDF). This would also get easier if `Value` gets changed back to the old type

polinavino replied:
It is hard to read, but i think it does not get us anything to use a type that's almost the real type used in the code, it only makes comparing the two more confusing

WhatisRT replied:
I think we discussed the definition of `Value` a lot in the past, and we settled on the old definition for the spec, and the optimized definition for the code. The main benefit of abstracting it like with this PR is that we can actually mention both in the spec. If we only wanted the code and the spec to agree, we could avoid all that and just could have changed the definition of `Value`. I think what we really should be doing is keeping the old definition of `Value`, spell out the isomorphism between it and the optimized version in the implementation and then say something along the lines of: all the functions for the optimized `Value` type of the implementation are derived via transporting them along that isomorphism.

WhatisRT commented on shelley-ma/formal-spec/shelley-ma.tex:
```diff
@@ -102,6 +102,8 @@
 \newcommand{\Ppm}{\type{Ppm}}
 \newcommand{\Value}{\type{Value}}
+\newcommand{\Integer}{\type{Integer}}
+\newcommand{\ValClass}{\mathcal{VAL}}
```
I don't like how this is typeset, multiple capital mathcal letters together often look weird. Maybe `\mathbf{Val}`?

WhatisRT committed:
Small fixes & changes from the audit

WhatisRT committed:
Spell checking

WhatisRT commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,77 +1,239 @@
+      & \text{Scale $\mathsf{ValType}$ value by integral constant}
+      \nextdef
+      %
+      & (-) ~\in~ \mathsf{ValType} \to \mathsf{ValType} \to \mathsf{ValType}\\
```
This definition is unused, so I think we should not have it

WhatisRT commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,77 +1,239 @@
+      & \text{Subtraction on $\mathsf{ValType}$}
+      \nextdef
+      %
+      & \fun{isZero} ~\in~ \mathsf{ValType} \to \Bool\\
```
Same as with `-`

WhatisRT commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,77 +1,239 @@
+      & \text{Check if a given $\mathsf{ValType}$ term contains only Ada}
+      \nextdef
+      %
+      & \fun{modifyCoin} ~\in~ (\Coin \to \Coin) \to (\mathsf{ValType} \to \mathsf{ValType})\\
```
Also unused

WhatisRT commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,77 +1,239 @@
+      & \text{Return the size in words of a $\mathsf{ValType}$ term}
+      \nextdef
+      %
+      & \fun{pointwise} ~\in~ (\Integer \to \Integer \to \Bool) \to (\mathsf{ValType} \to \mathsf{ValType} \to \Bool)\\
```
Also unused

polinavino replied:
it is technically used where we use "<=" on Values. I am not sure if I should make a note of that or replace the math symbol with `pointwise`

WhatisRT replied:
I think we want to put the requirement of a partial order on the monoid aspect. I.e. instead of saying we want a monoid with additional functions, we want a partially ordered monoid with additional functions. In fact, a monoid with the module structure given by `*` is already a group, and we probably want it to be commutative. So, I guess we want to say `partially ordered abelian group with extra stuff`. Then we can also drop a huge bunch of stuff in this figure (`zero, +, *, -, <=`)

WhatisRT replied:
Also, partially ordered abelian group also implies monotonicity of functions like `+`, which I think is a good thing to require

polinavino replied:
I dont think we should drop any of this, since we do actually need to specify it, and it might be better not to assume that people know these things are implied by partially ordered abelian group structure

WhatisRT replied:
Instead of putting it in the figure, there could be a sentence or short paragraph reminding the reader what a partially ordered abelian group is

WhatisRT replied:
So the thing with `pointwise` is that it's used to define the relations of a specific `ValMonoid` (i.e. `Value`). `ValMonoid` should not require `pointwise` to be defined, but it should require `<=` etc. to be defined.

polinavino replied:
i do not agree with this - i think using pointwise makes the spec more clear and easier to compare to the code

WhatisRT replied:
It's a fault of the code (inherited from Haskell) though, that it isn't using things like `<=` to compare. If I'd implement the spec in another language, there's a very high likelihood that it would be using `<=` and not `pointwise (<=)`. So the code should maybe mention in a comment why it isn't using `<=` to explain this deviation.

WhatisRT commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,77 +1,239 @@
+the size of the serialized representation of a term of the underlying type.
+This function is used in the UTXO rule.
+We do not include it in the monoid definition because it cannot be defined
+explicity separately from concrete implementation.
```
This should actually be part of the definition of `ValMonoid`. If we were to define it, it would have different implementations for each type

polinavino replied:
I think it should not because
- we cant/wont actually define this function in the spec because it is a Haskell function that is opaque to the ledger
- even if we dig up its specification and give it here, it is misleading to pretend that this spec of `serSize` is defined as part of specifying a `ValMonoid` rather than exists for all serializable types

polinavino replied:
changed it to the computation of a serialized representation

WhatisRT commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,77 +1,239 @@
+they return a $\ValMonoid$ containing only the Ada tokens contained in the original $\ValMonoid$.
+
+\textbf{Fungible and non-fungible tokens.}
+Recall that a token is said to be fungible at a specific time (slot $s$, for example) if
```
I don't think we should mention time here, it has nothing to do with fungibility in our ledger

polinavino replied:
That's how people are making NFTs now - by making them with a policy that is invalid after a given slot. And usually people mean that an NFT is an NFT forever. Whereas we mean it in the weaker sense "no other ones with the same asset ID exist in some ledger state. I think it is useful to disambiguate here

WhatisRT replied:
I would actually be quite happy with not having this fungibility discussion in the spec at all. Also, I don't think I want to define what a `token` is because it seems to be confusing and not particularly helpful, especially since the definition is not relevant to the ledger at all.

WhatisRT commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,77 +1,239 @@
     \begin{array}{r@{~\in~}l@{\qquad=\qquad}lr}
+      \var{aname} & ~\AssetName & \mathsf{ByteString} \\
       \var{pid} & \PolicyID & \ScriptHash \\
+      \var{adaID} & ~\AdaIDType & \{~\mathsf{AdaID}~\} \\
```
This type is unused

polinavino replied:
it is unused because it does not exist the ledger implementation. It is used in the explanation of how to implement Ada with this example of `ValMonoid`

WhatisRT replied:
The only places where the string `AdaIDType` exist in the source files are here, and at the definition of the `AdaIDType` macro. So this type is currently not mentioned anywhere except here

polinavino replied:
i added a brief description of it

WhatisRT replied:
I understand now. It's still unused though, and to fix that, you have to replace `AdaID` in the line below by `AdaIDType`

lehins committed:
Fix spelling mistake in spec

WhatisRT commented on shelley-ma/formal-spec/transactions.tex:
```diff
@@ -109,18 +109,18 @@ \subsection*{The Mint Field}
 The body of a transaction in the ShelleyMA era contains one additional
 field, the $\fun{mint}$ field.
 The $\fun{mint}$ field is a term of type $\ValMonoid$, which contains
```
'term of type' is a meta-level construction that I don't think we should be using, because it has some potential connotations that we don't want. For example, as terms, `0 + 0 \neq 0`. We could just say 'The mint field is of type \ValMonoid`

WhatisRT commented on shelley-ma/formal-spec/utxo.tex:
```diff
@@ -80,25 +87,27 @@ \subsection*{UTxO Helper Functions}
+If a transaction $\var{tx}$ has an empty $\fun{mint}$ field, the preservation
 of value reduces to an equation that the sum of inputs and the sum of
 outputs are equal, which is exactly the same condition as for Shelley,
 except that there are no administrative fields. If a transactions
```
It's not clear to me what 'except that there are no administrative fields' means

WhatisRT replied:
Also, typo: transactions -> transaction

WhatisRT commented on shelley-ma/formal-spec/valmonoid-examples.tex:
```diff
@@ -2,23 +2,26 @@ \section{Other $\ValMonoid$ Examples}
-that Ada indeed has a minting policy, and could potentially have several
-asset names scoped under this policy. \\~\\
+that Ada indeed has a minting policy, $\mathsf{AdaPolicyID}$. It must be specified in
+order to implementat of this version of asset bundles, but we leave abstract here.
```
Typo: implementat

WhatisRT replied:
Also, 'but we leave abstract here' -> add an 'it'

WhatisRT commented on shelley-ma/formal-spec/value-size.tex:
```diff
@@ -4,7 +4,7 @@ \section{Output Size}
 disincentivize attacking nodes by putting many small outputs on the
 UTxO, permanently blocking memory. In the Mary era
-and onwards, the $\ValMonoid$ type replaces $\Coin$ in UTxO map. $\ValMonoid$ terms
+and onwards, the $\ValMonoid$ is $\Value$, rather than $\Coin$ (as in Allegra). $\Value$ terms
```
Again, 'terms' isn't good, maybe 'elements of Value'

WhatisRT commented on shelley-ma/formal-spec/value-size.tex:
```diff
@@ -143,23 +149,22 @@ \subsection{Min UTxO Value}
-the size of a UTxO entry, not counting the size of the $\Value_C$ term it contains.
-Here, again, the actual size of a UTxO (excluding the $\Value_C$) can vary, but
+the size of a UTxO entry, not counting the size of the $\ValMonoid$ term it contains.
+Here, again, the actual size of a UTxO (excluding the $\ValMonoid$ term) can vary, but
```
Another 'term', twice

WhatisRT commented on shelley-ma/formal-spec/utxo.tex:
```diff
@@ -73,33 +80,34 @@ \subsection*{UTxO Helper Functions}
 
   \item The $\fun{consumed}$ and $\fun{produced}$ calculations are similar to their Shelley
-    counterparts, with the following changes: 1) They return elements of $\Value$, which
+    counterparts, with the following changes: 1) They return terms of $\ValMonoid$, which
```
'elements' is better than 'terms'

WhatisRT commented on shelley-ma/formal-spec/transactions.tex:
```diff
@@ -102,32 +106,69 @@ \subsection*{New Script Type}
 
+Note that the $\fun{mint}$ field exists in both Allegra and Mary eras,
+but can only be used in the Mary era, when multi-assets are introduced.
+In Figure \ref{fig:minted} we give two versions of a function $\fun{minted}$,
```
During our audit, Alex and I came across `minted` and we really didn't like it. We'd have removed it if we had the time to do so before the HF, but my guess would be that it'll be removed later. So I really don't want it to creep into the spec as well.

WhatisRT replied:
The correct way to go about this is to add a function like `policies` to the definition of `ValMonoid`.

WhatisRT replied:
I think you've changed it to be like that, but the actual entry for `policies` is missing in the figure for `ValMonoid`

polinavino replied:
right, oops

WhatisRT commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,125 +1,289 @@
+\ref{fig:coin}. Addition and multiplication are defined in the usual way
+(as on $\Integer$).
+
+When multi-asset support on the ledger is introduced, Ada ($\Coin$) is still expected to be
```
I'm not sure what the purpose of this sentence is. Also, most common by what metric?

polinavino replied:
by the metric that every single output contains some

WhatisRT replied:
If this is the intended meaning of this sentence, I think 'expected' is not the correct word, since this is enforced by the ledger. Maybe 'will' instead?

lehins committed:
Implement Generation of Certificates. Bug repro with '-p UTXOW --quickcheck-replay=565274'

WhatisRT committed:
Merge branch 'master' of github.com:input-output-hk/cardano-ledger-specs into andre/alonzo-audit

lehins committed:
Finish work on UTXOW generators:

* Utilize LEDGER rules for UTXOW testing

* Workaround a misfeature where duplicate scripts in DCert are ignored

* Get generation of rewards working

* Generation of Delegation certificates

* Occasionally reuse existing credentials

WhatisRT committed:
Add some missing information

WhatisRT committed:
Fix arrow for bootstrap address size checks

WhatisRT commented on shelley-ma/formal-spec/transactions.tex:
```diff
@@ -69,28 +68,33 @@
-native, meaning
-that they are evaluated by the ledger directly. We specify the evaluation
-function for the additional script type in section~\ref{sec:timelock-lang}.
+$\Timelock$. We specify the evaluation
```
I think it should be mentioned here that it's backwards compatible to multi-sig

polinavino replied:
that is described in the transaction translation section

WhatisRT replied:
I see. I don't think that's a good place for that though, it's not very discoverable. This is the place where the new script type is explained, so it's in a much more visible location. Maybe at least mention here that it is compatible, and link to the other section for how?

polinavino replied:
makes sense!

WhatisRT commented on shelley-ma/formal-spec/utxo.tex:
```diff
@@ -7,7 +7,7 @@ \subsection*{UTxO Helper Functions}
   \begin{equation*}
     \begin{array}{r@{~\in~}l@{\qquad=\qquad}lr}
-      \var{mest} & \MemoryEstimate & \N
+      \var{mest} & \MemoryEstimate & \Integer
```
Is there a good reason for this to be an Integer, or is this just to sync with the code? Because I'd say that negative numbers don't make sense for the memory estimate, and it's the code that should be changed here

polinavino replied:
i changed it because auditors pointed this out

WhatisRT replied:
Yeah, but it was pointed out that this is a difference between spec and code, and I'd say the code is wrong in this case

polinavino replied:
i think we should keep them aligned, but if this will prevent merginig this PR in the near future, feel free to change

WhatisRT replied:
I also want things to be aligned, but they should be aligned to what should be correct, instead of to something that just happened to end up in the code

WhatisRT commented on shelley-ma/formal-spec/utxo.tex:
```diff
@@ -170,11 +182,10 @@ \subsection*{The UTXO Transition Rule}
-      \hldiff{\fun{getValue}~txout \geq} \\
-      \hldiff{\fun{inject}~(\fun{max}~(\var{minUTxOValue}~pp,~\fun{utxoEntrySize}~{txout} * \fun{adaPerUTxOWord}~pp))} \\~
+      \hldiff{\forall txout \in \txouts{txb},~\fun{getValue}~txout \geq} \\
+      \hldiff{\fun{inject}~(\fun{scaledMinDeposit}~v~(\fun{minUTxOValue}~pp))} \\~
```
I think there shouldn't be a line break on the inequality sign (I think one of the reviewers in the past got confused because they thought something was missing)

lehins committed:
HLint for shelley-ma impl is done

lehins committed:
HLint Shelley MA tests are done too.

lehins committed:
Remove usage of unsafePerformIO and get rid of dead code

lehins committed:
Refactor reusable functions and reduce number of long wraparound lines

lehins committed:
Few minor improvements and a fix for 'Reward Provenance' property test:

* Faulty test was caused by the `genBlocksMade` producing an empty map
  sometimes. Fixed
* Turn printer into generator
* Get rid of non-reproducable `generate`
* Consolidate `reward` and `rewardPulser` functions into one reducing clutter.

WhatisRT committed:
Introduce the notion of a Token algebra

WhatisRT committed:
Introduce the notion of a Token algebra

Commit:
Fix transaction cddl to include IsValid

polinavino commented on shelley-ma/formal-spec/utxo.tex:
```diff
@@ -33,6 +33,9 @@ \subsection*{UTxO Helper Functions}
     & \fun{ubalance} ~ utxo = \hldiff{\sum_{\wcard\mapsto\var{u}\in~\var{utxo}} \fun{getValue}~\var{u}}
     \nextdef
+    & \fun{adaOnly} ~\in~ \mathsf{ValType} \to \Bool \\
+    & \fun{adaOnly}~v = v \in \range \fun{inject}
```
leave this as composition of coin and inject - imo the goal is to make this as easy to read and understand as possible

polinavino commented on shelley-ma/formal-spec/transactions.tex:
```diff
@@ -83,19 +83,16 @@ \subsection*{New Output Type}
 
-In the Allegra era, the $\ValMonoid$ is $\Coin$, so there is no
-multi-asset support in the UTxO or the transactions. In the Mary era, it
-is changed to $\Value$ in order to add multi-asset support.
```
I would prefer to leave this in

polinavino commented on shelley-ma/formal-spec/valmonoid-examples.tex:
```diff
@@ -43,24 +43,21 @@ \subsection{No Special Case for Ada}
+    & \fun{inject}~c  = \mathsf{AdaPolicyID} \mapsto_0 (\mathsf{AdaName} \mapsto_0 c)
     \nextdef
     %
-    & \fun{zero} ~=~ \Nothing
```
we want to give all the functions that are in the token algebra structure, which include zero

polinavino commented on shelley-ma/formal-spec/valmonoid-examples.tex:
```diff
@@ -90,50 +86,44 @@ \subsection{Special Case Ada Representation}
-    & \fun{anames}~(c, ~v) ~=~ \bigcup_{pid \in\fun{policies}{~v}} \supp{(v~pid)}
-    \nextdef
-    %
-    & \fun{zero} ~=~ (0, \Nothing)
```
we should keep zero

polinavino commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,213 +1,156 @@
+  partial order, such that addition is monotonic in both variables)
+  together with the functions and properties as described in Figure
+  \ref{fig:TokenAlgebra}.
+\end{definition}
```
NICE

polinavino commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -272,26 +223,11 @@ \subsection{Multi-assets and $\Value$ in the Mary Era}
-\subsection{Pointwise Value Operations and Partial Order}
+\subsection{Fixing a Token algebra}
 
-We have a number of operations on $\Value$ left to define for it to have
```
I really, really do not agree with removing explicit definitions of pointwise scalar multiplication and comparisons. We should not expect people to know how this works.

polinavino commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,213 +1,156 @@
+    & \fun{inject}~c  = \mathsf{AdaID}~\mapsto_0~\var{c}
     \nextdef
     %
-    & \fun{zero} ~=~ \Nothing
```
zero should stay

polinavino commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,213 +1,156 @@
   \begin{align*}
-      & \fun{policies} ~=~ \Nothing
-      \nextdef
-      & \fun{zero} ~=~ 0
```
zero should stay

polinavino commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,213 +1,156 @@
+  \emph{Functions}
+  %
   \begin{align*}
-      & \mathsf{ValType} ~\in~ \mathsf{Type}  \\
```
we should keep a full list of the functions so that a reader not familiar with algebraic structures does not need to look it up

polinavino commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,213 +1,156 @@
+\begin{itemize}
+  \item the Allegra ledger rules with $\Coin$, and
+  \item the Mary ledger rules with $\Value$ (defined below).
+\end{itemize}
```
this should not go in the Coin/Allegra subsection, but in the intro

polinavino commented on shelley-ma/formal-spec/value.tex:
```diff
@@ -1,213 +1,156 @@
+In this chapter we introduce the concept of a \emph{Token algebra},
+which is an abstraction used to generalize this specification to two
+different eras depending on the Token algebra chosen.
 
```
we should remind the reader here the purpose of introducting this token algebra thing - to represent different types of collections of assets (coins or heterogeneous collections)

nc6 committed:
Merge pull request #2455 from input-output-hk/lehins/hlint-shelley-ma

Apply hlint suggestions and other cleanup to shelley-ma

nc6 committed:
Add a test that events are sufficiently lazy.

We want to ensure that events are not prematurely evaluated, even when
using the `EPReturn` policy.

polinavino committed:
Merge pull request #2448 from input-output-hk/andre/alonzo-audit

Alonzo audit

Commit:
Revert "Refactored RewardSnapShot, and added Pulsing to SnapShots."

This reverts commit a7d4911a1b47bfc5c946a7e821ea5a3a3dc1e77a.

Commit:
remove redundant imports

lehins committed:
Rename `lib` -> `src`

lehins committed:
Rename `alonzo/test` -> `alonzo/test-suite`

lehins committed:
Rename `shelley-ma/shelley-ma-test` -> `shelley-ma/test-suite`

lehins committed:
Move `plutus-preprocessor` -> `libraries/plutus-preprocessor`

lehins committed:
Remove `cardano-prelude` from `cardano-ledger-shelley-ma-test` dependencies

lehins committed:
Move `semantics/executable-spec` -> `libraries/small-steps`, `semantics/small-step-test` -> `libraries/small-step-test`

lehins committed:
Move `example-shelley` -> `libraries/cardano-ledger-example-shelley`

lehins committed:
Move `shelley/chain-and-ledger/executable-spec` -> `shelley/impl` and `shelley/chain-and-ledger/shelley-spec-ledger-test` -> shelley/test-suite`

lehins committed:
Move `shelley/chain-and-ledger/dependencies/non-integer` -> `libraries/shelley-spec-non-integral`.

Also:

* move `shelley/chain-and-ledger/dependencies/non-integer/doc` ->
       `docs/non-integer-calculations`
* Remove stack related stuff

lehins committed:
Move top level libs to `libraries`:

* Move `cardano-ledger-core` -> `libraries/cardano-ledger-core`

* Move `cardano-ledger-test` -> `libraries/cardano-ledger-test`

* Move `cardano-protocol-tpraos` -> `libraries/cardano-protocol-tpraos`

nc6 committed:
Merge pull request #2481 from input-output-hk/nc/test-lazy-events

Add a test that events are sufficiently lazy.

Commit:
Merge pull request #2484 from input-output-hk/jc/benchmark/1-30-0-rc3-without-pulse

Revert stake distribution pulser

goolord committed:
RetiredPools event

nc6 committed:
Bump cardano base

goolord committed:
associate accouns with rewards

goolord committed:
fix instances

goolord committed:
use unionWith (<>)

Commit:
Merge pull request #2488 from input-output-hk/nc/bump-base

Bump cardano base

Commit:
Merge pull request #2487 from input-output-hk/zach/poolreap-events

RetiredPools event

Commit:
Merge pull request #2434 from input-output-hk/lehins/alonzo-utxow-props

Alonzo utxow Tx generator and properties

lehins committed:
Move `pool-ranking` and `semantics` specs into their own folder `docs`:

* Move `shelley/pool-ranking` -> `docs/pool-ranking`
* Move `semantics/formal-spec` -> `docs/small-step-semantics`

lehins committed:
Move `shelley/chain-and-ledger/formal-spec` -> `shelley/formal-spec`

Commit:
Merge pull request #2456 from input-output-hk/jc/fix-tx-cddl

Fix transaction cddl to include IsValid

lehins committed:
Move all era specific projects into `./eras/`

lehins committed:
Rename `libraries` -> `libs`

lehins committed:
Rename all modules `Shelley.Spec.Ledger.*` -> `Cardano.Ledger.Shelley.*`:

Rename also applies to `shelley-spec-ledger-test`:
`Test.Shelley.Spec.Ledger.*` -> `Test.Cardano.Ledger.Shelley.*`:

Rename `STS` -> `Rules` for consistency:
`Shelley.Spec.Ledger.STS` -> `Cardano.Ledger.Shelley.Rules`

lehins committed:
Rename packages `shelley-spec-ledger` and `shelley-spec-ledger-test`:

* `shelley-spec-ledger` -> `cardano-ledger-shelley`
* `shelley-spec-ledger-test` -> `cardano-ledger-shelley-test`

lehins committed:
Rename package: `shelley-spec-non-integral` -> `non-integral`

lehins committed:
Define deprecated `shelley-spec-ledger` and `shelley-spec-ledger-test` to avoid breakage

lehins committed:
Create backwards compatible `shelley-spec-non-integral` -> `non-integral`:

* Rename `Shelley.Spec.NonIntegral` -> `Cardano.Ledger.NonIntegral`
* Create deprecated `shelley-spec-non-integral` that reexports the
  new `Cradano.Ledger.NonIntegral` module from `non-integral` package.

wolverineks committed:
fix bullet points

wolverineks committed:
fix script example

wolverineks committed:
fix typo

wolverineks committed:
fix typo

lehins committed:
Fixup rnames in hie, script and tex files

lehins committed:
Fix relative paths

lehins committed:
Fix relative dirs for byron spec nix derivations

lehins committed:
Create symlinks for:

    * `alonzo` -> `eras/alonzo`
    * `alonzo/test` -> `eras/alonzo/test-suite`
    * `byron` -> `eras/byron`
    * `shelley` -> `eras/shelley`
    * `shelley/pool-ranking` -> `docs/pool-ranking`
    * `shelley/chain-and-ledger/formal-spec` -> `eras/shelley/formal-spec`
    * `shelley/chain-and-ledger/dependencies/non-integer/doc` -> `docs/non-integer-calculations`
    * `shelley-ma` -> `eras/shelley-ma`
    * `shelley-ma/shelley-ma-test` -> `eras/shelley-ma/test-suite`
    * `semantics/executable-spec` -> `libs/small-steps`
    * `semantics/formal-spec` -> `docs/small-step-semantics`
    * `semantics/small-steps-test` -> `libs/small-steps-test`
    * `cardano-ledger-core` -> `libs/cardano-ledger-core`
    * `cardano-ledger-test` -> `libs/cardano-ledger-test`
    * `cardano-protocol-tpraos` -> `libs/cardano-protocol-tpraos`
    * `example-shelley` -> `libs/cardano-ledger-example-shelley`
    * `plutus-preprocessor` -> `libs/plutus-preprocessor`

lehins committed:
Update README to reflect new paths

goolord committed:
add `epochNo` to `RetiredPools`

Commit:
Merge pull request #2495 from input-output-hk/zach/epochno-retiredpools

add `epochNo` to `RetiredPools`

lehins committed:
Merge pull request #2497 from input-output-hk/lehins/investigate-test-failure

Few minor improvements and a fix for 'Reward Provenance' property test:

lehins committed:
Merge pull request #2483 from input-output-hk/lehins/organize-directories-and-modules

Organize locations for Haskell packages

lehins committed:
Remove `UNPACK` pragma that has no affect

lehins committed:
Remove unused dependencies

lehins committed:
Disable unimportant warnings in byron code

lehins committed:
Get rid of `development` cabal flag in favor of `cabal configure --ghc-options="-Wwarn"`

lehins committed:
Stream tests output. Ensure tests and benchmarks are always build.

lehins committed:
Fix nightly builds

lehins committed:
Fix module duplication in library, tests and benchmarks

lehins committed:
Add a bit more consistency to cabal files

Commit:
remove check-cabal-project pipeline

lehins committed:
Merge pull request #2498 from input-output-hk/lehins/warnings-and-warts

Warnings and warts

Commit:
Merge pull request #2492 from wolverineks/master

fix typos

Commit:
fix alonzo CDDL link in the readme

Commit:
Merge pull request #2499 from input-output-hk/jc/gh-ormolu-action

use github action to run ormolu

Commit:
Merge pull request #2500 from input-output-hk/jc/fix-alonzo-cddl-link

fix alonzo CDDL link in the readme

WhatisRT committed:
Merge branch 'polina/formal-fix2' of github.com:input-output-hk/cardano-ledger-specs into polina/formal-fix2
